## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [1]:
import pandas as pd
df = pd.read_csv('Resources/client_dataset.csv')
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [2]:
# View the column names in the data
df.columns 

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [3]:
# Use the describe function to gather some basic statistics
basic_stats = df.describe()
basic_stats

client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [4]:
# Use this space to do any additional research
# and familiarize yourself with the data.
df.count()

first          54639
last           54639
job            54639
phone          54639
email          54639
client_id      54639
order_id       54639
order_date     54639
order_week     54639
order_year     54639
item_id        54639
category       54639
subcategory    54639
unit_price     54639
unit_cost      54639
unit_weight    54639
qty            54639
line_number    54639
dtype: int64

In [5]:
df.values

array([['Donald', 'Harding', 'Immunologist', ..., 7.5, 105, 1],
       ['Tiffany', 'Myers', 'Music therapist', ..., 1.49, 21, 0],
       ['Shannon', 'Watson', 'Immunologist', ..., 1.68, 39, 6],
       ...,
       ['Monica', 'Gutierrez', 'Graphic designer', ..., 4.16, 475, 6],
       ['Wanda', 'Solomon', 'Toxicologist', ..., 18.04, 112, 0],
       ['Dwayne', 'Bell', 'Chief Financial Officer', ..., 2.07, 1031, 5]],
      dtype=object)

In [6]:
df.dtypes

first           object
last            object
job             object
phone           object
email           object
client_id        int64
order_id         int64
order_date      object
order_week       int64
order_year       int64
item_id         object
category        object
subcategory     object
unit_price     float64
unit_cost      float64
unit_weight    float64
qty              int64
line_number      int64
dtype: object

In [7]:
# What three item categories had the most entries?
cat_df = df.sort_values("category")

sort_cat_df = cat_df['category'].value_counts()

sort_cat_df.head(3)

# Consumables,furniture and software are the 3 categories with the most entries.

category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [8]:
# For the category with the most entries, which subcategory had the most entries?
sub_df = df.loc[df['category'] == 'consumables']

sub_df['subcategory'].value_counts().head(1)


subcategory
bathroom supplies    6424
Name: count, dtype: int64

In [9]:
# Which five clients had the most entries in the data?

top = df['client_id'].value_counts()

top.head(5)

client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

In [10]:
# Store the client ids of those top 5 clients in a list.

top_list = []
for (client_id) in top.head(5).index:
    top_list.append(client_id)
    
top_list



[33615, 66037, 46820, 38378, 24741]

In [11]:
# How many total units (the qty column) did the client with the most entries order order?

top_units = top_list[0]

total_units = df[df['client_id'] == top_units]['qty'].sum()

total_units



64313

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [12]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty
 
subtot = df["unit_price"] * df["qty"] 
df["subtotal"] = subtot

subtotal_df = df[["unit_price", "qty", "subtotal"]]

subtotal_df.head()


unit_price  qty   subtotal
0     1096.80  105  115164.00
1       24.95   21     523.95
2       13.52   39     527.28
3       36.42   29    1056.18
4      195.10   20    3902.00

In [13]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.

df["tot_weight"] = df["unit_weight"] * df["qty"]

df["ship_cost"] = df["tot_weight"].apply(lambda x: 7* x if x >50 else 10*x)

shipping_df= df

shipping_df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ...     category         subcategory unit_price  \
0          17        2023  ...        decor            wall art    1096.80   
1          20        2023  ...  consumables                pens      24.95   
2           4        2023  ...     software  project management      13.52   
3          17        2023  ...  consumables                pens      36.42   
4          21        2023  ...  consumables                misc     195.10   

   unit_cost  unit_weight  qty  line_number   subtotal  tot_weight  ship_cost  
0     762.71         7.50  105            1  115164.00      787.50    5512.50  
1      15.09         1.49   21            0     523.95       31.29     312.90  
2       7.86         1.68   39            6     527.28       65.52     458.64  
3      24.85         1.23   29            3    1056.18       35.67     356.70  
4     108.17        46.43   20            1    3902.00      928.60    6500.20  

[5 rows x 21 columns]

In [14]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%

totp = (df["subtotal"] + df["ship_cost"]) * 1.0925

df["tot_price"] = totp.round(2)

total_df = df

total_df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ...         subcategory unit_price unit_cost  \
0          17        2023  ...            wall art    1096.80    762.71   
1          20        2023  ...                pens      24.95     15.09   
2           4        2023  ...  project management      13.52      7.86   
3          17        2023  ...                pens      36.42     24.85   
4          21        2023  ...                misc     195.10    108.17   

   unit_weight  qty  line_number   subtotal  tot_weight  ship_cost  tot_price  
0         7.50  105            1  115164.00      787.50    5512.50  131839.08  
1         1.49   21            0     523.95       31.29     312.90     914.26  
2         1.68   39            6     527.28       65.52     458.64    1077.12  
3         1.23   29            3    1056.18       35.67     356.70    1543.57  
4        46.43   20            1    3902.00      928.60    6500.20   11364.40  

[5 rows x 22 columns]

In [15]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).

totc = df["ship_cost"] + (df["unit_cost"] * df["qty"])

df['tot_cost'] = totc

df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_price unit_cost unit_weight  qty  \
0          17        2023  ...    1096.80    762.71        7.50  105   
1          20        2023  ...      24.95     15.09        1.49   21   
2           4        2023  ...      13.52      7.86        1.68   39   
3          17        2023  ...      36.42     24.85        1.23   29   
4          21        2023  ...     195.10    108.17       46.43   20   

   line_number   subtotal  tot_weight  ship_cost  tot_price  tot_cost  
0            1  115164.00      787.50    5512.50  131839.08  85597.05  
1            0     523.95       31.29     312.90     914.26    629.79  
2            6     527.28       65.52     458.64    1077.12    765.18  
3            3    1056.18       35.67     356.70    1543.57   1077.35  
4            1    3902.00      928.60    6500.20   11364.40   8663.60  

[5 rows x 23 columns]

In [16]:
# Create a column for the profit of each line using line cost and line price

prof = (df["tot_price"] - df["tot_cost"]) 

df["profit"] = prof

total_df = df

total_df.head()


first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_cost unit_weight  qty  line_number  \
0          17        2023  ...    762.71        7.50  105            1   
1          20        2023  ...     15.09        1.49   21            0   
2           4        2023  ...      7.86        1.68   39            6   
3          17        2023  ...     24.85        1.23   29            3   
4          21        2023  ...    108.17       46.43   20            1   

    subtotal  tot_weight  ship_cost  tot_price  tot_cost    profit  
0  115164.00      787.50    5512.50  131839.08  85597.05  46242.03  
1     523.95       31.29     312.90     914.26    629.79    284.47  
2     527.28       65.52     458.64    1077.12    765.18    311.94  
3    1056.18       35.67     356.70    1543.57   1077.35    466.22  
4    3902.00      928.60    6500.20   11364.40   8663.60   2700.80  

[5 rows x 24 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [17]:
# Check your work using the totals above 

# set order number
target_order = 2742071

# sum total price for the target Order ID
order_details = df[df['order_id'] == target_order]
order_sum = order_details['tot_price'].sum()

# Set the target total price
target_total = 152811.89

# print the result
if order_sum == target_total:
    print(f"The total price for Order ID 2742071 confirmed as: Calculated: ${order_sum:.2f}, Expected: ${target_total:.2f}.")
else:
    print(f"The total price for Order ID 2742071 does not match. Calculated: ${order_sum:.2f}, Expected: ${target_total:.2f}.")

The total price for Order ID 2742071 does not match. Calculated: $152811.89, Expected: $152811.89.


In [18]:
# Check your work using the totals above
# set order number
target_order = 2173913

# sum total price for the target Order ID
order_details = df[df['order_id'] == target_order]
order_sum = order_details['tot_price'].sum()

# Set the target total price
target_total = 162388.71

# print the result
if order_sum == target_total:
    print(f"The total price for Order ID 2742071 confirmed as: Calculated: ${order_sum:.2f}, Expected: ${target_total:.2f}.")
else:
    print(f"The total price for Order ID 2742071 does not match. Calculated: ${order_sum:.2f}, Expected: ${target_total:.2f}.")

The total price for Order ID 2742071 confirmed as: Calculated: $162388.71, Expected: $162388.71.


In [19]:
# Check your work using the totals above
new_ord_id = 6128929

# Summing the total price for the specific Order ID
order_details = df[df['order_id'] == new_ord_id]
order_sum = order_details['tot_price'].sum()

# Set the target total price
target_total = 923441.25

# print the result
if order_sum == target_total:
    print(f"The total price for Order ID 2742071 confirmed as: Calculated: ${order_sum:.2f}, Expected: ${target_total:.2f}.")
else:
    print(f"The total price for Order ID 2742071 does not match. Calculated: ${order_sum:.2f}, Expected: ${target_total:.2f}.")

The total price for Order ID 2742071 confirmed as: Calculated: $923441.25, Expected: $923441.25.


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [20]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids. FIX - add $ sign
  
for client_id in top_list:
    amount_spent = total_df[total_df['client_id'] == client_id]['tot_cost'].sum()
    print(f"{client_id}: ${amount_spent:,.2f}")


33615: $6,175,313.91
66037: $7,004,482.98
46820: $7,007,191.64
38378: $9,634,720.98
24741: $45,688,899.71


In [21]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit. 

summary_df = total_df[total_df['client_id'].isin(top_list)].groupby('client_id').agg(
    total_units_purchased=('qty', 'sum'),
    total_shipping_price=('ship_cost', 'sum'),
    total_revenue=('tot_cost', 'sum'),
    total_profit=('profit', 'sum')
)
new_df = pd.DataFrame(summary_df)

print(new_df.head()) 

           total_units_purchased  total_shipping_price  total_revenue  \
client_id                                                               
24741                     239862            5126448.37    45688899.71   
33615                      64313            1828984.89     6175313.91   
38378                      73667            3429455.40     9634720.98   
46820                      75768            1601448.84     7007191.64   
66037                      43018            1395151.85     7004482.98   

           total_profit  
client_id                
24741       36579992.31  
33615        2201994.61  
38378        3271829.89  
46820        2736602.72  
66037        3255031.81  


In [26]:
# Format the data and rename the columns to names suitable for presentation.  Fix - currency
# Rename columns
summary_df = summary_df.rename(columns={
    'client_id': 'Client ID',
    'total_units_purchased': 'Total Purchased',
    'total_shipping_price': 'Total Shipping',
    'total_revenue': 'Total Revenue',
    'total_profit': 'Total Profit',
})

# Format columns
summary_df['Total Purchased'] = summary_df['Total Purchased'].apply(lambda x: "${:,.2f}".format(x))
summary_df['Total Shipping'] = summary_df['Total Shipping'].apply(lambda x: "${:,.2f}".format(x))
summary_df['Total Revenue'] = summary_df['Total Revenue'].apply(lambda x: "${:,.2f}".format(x))
summary_df['Total Profit'] = summary_df['Total Profit'].apply(lambda x: "${:,.2f}".format(x))

money_columns = ['Total Shipping', 'Total Revenue', 'Total Profit']

# Define a function that converts a dollar amount to millions.

def dollars_to_millions(amount):
    return amount / 1_000_000

# Apply the currency_format_millions function to only the money columns. 
summary_df[money_columns].apply(dollars_to_millions)

# Rename the columns to reflect the change in the money format. 

summary_df = summary_df.rename(columns={
    'Total Purchased': 'Total Purchased (millions)',
    'Total Shipping': 'Total Shipping (millions)',
    'Total Revenue': 'Total Revenue (millions)',
    'Total Profit': 'Total Profit (millions)',
})

summary_index_df

TypeError: unsupported operand type(s) for /: 'str' and 'int'

In [ ]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.
summary_million = summary_index_df.sort_values(by='Total Profit (millions)', ascending=False)
summary_million
